src/osyris/spatial/spatial.py

```diff
@@ -0,0 +1,89 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Osyris contributors (https://github.com/nvaytet/osyris)
+
+import numpy as np
+from ..core import Array
+from ..config import units
+
+
+def rotation_matrix(vec, angle):
+    """
+    Returns 3D rotation matrix of angle 'angle' around rotation vector 'vec'.
+    """
+    if isinstance(vec, list):
+        vec = np.array(vec)
+    vec = vec / np.linalg.norm(vec)
+    R = np.cos(angle) * np.identity(3) + (np.sin(angle)) * np.cross(
+        vec,
+        np.identity(vec.shape[0]) * -1) + (1 - np.cos(angle)) * (np.outer(vec, vec))
+    return R
+
+
+def spherical_to_cartesian(r, theta, phi):
+    """
+    Converts spherical components radius, colatitude and azimuth to x,y,z
+    Returns an osyris array of shape (len(r), 3) containing a stacked xyz
+    """
+    x, y, z = (r * np.cos(phi) * np.sin(theta), r * np.sin(phi) * np.sin(theta),
+               r * np.cos(theta))
+    xyz = np.vstack([x, y, z])
+    return Array(values=np.transpose(xyz).values, unit=r.unit, name="position")
+
+
+def cartesian_to_spherical(position, origin=[0, 0, 0]):
```

Comment:
In your original suggestion, you also had vector bases as part of the arguments. I think these would still be needed here?

Reply:
Indeed, I just wanted to keep it simple and generic for now, but expressing your coordinates in a new basis requires you to simply rotate them (which is the purpose of the `rotation_matrix` function).

Reply:
Maybe thinking about how one would use this in a notebook would be a good way to figure out the API.

You have a disk somewhere in your simulation. You want to basically rotate everything so that the z axis is aligned with the angular momentum (is that all you need to do?)

You could have the following steps:
- find angular momentum vector
- rotate all coords and vectors into a new basis (still in cartesian coords)
- then access the radial or angular components of your vectors (this could be done via the `.r` and `.theta` properties?

The problem with just having `.r` property on a vector is that it works for position vectors, but not for other vectors such as velocity. If you want the radial velocity, you need to take the scalar product of the velocity vector with the position vector, so the conversion needs both position vectors and velocity vectors as inputs.

Maybe that is an argument for having functions that turn a whole `Datagroup` into spherical or cylindrical components?
Something like
```
spherical_hydro = osyris.spatial.get_spherical_components(data['hydro'])
```
Which would return a new Datagroup (i think it might be safer to return a new group rather than doing operations in-place).

But then the `get_spherical_components` also needs the `data['amr']['position']` array...
Not sure if sending the entire Dataset to be converted is good, because you may be doubling your RAM usage.

Maybe the function should just accept a single Array, and an optional `positions` argument, which could just be `None` for computing the spherical components of the cell positions?

Reply:
> You have a disk somewhere in your simulation. You want to basically rotate everything so that the z axis is aligned with the angular momentum (is that all you need to do?)

Yes, that's all there is to it.

> The problem with just having .r property on a vector is that it works for position vectors, but not for other vectors such as velocity. If you want the radial velocity, you need to take the scalar product of the velocity vector with the position vector, so the conversion needs both position vectors and velocity vectors as inputs.

Yeah that's what's been causing me a bit of a headache. It's difficult to come up with a clean API for something like this.

> Not sure if sending the entire Dataset to be converted is good, because you may be doubling your RAM usage.

That is why in #52 I suggested that we make the coordinate system an attribute of the Dataset. That way with `data.transform(system="spherical", basis="top")`, you do not return a new dataset, but simply convert all components in it to spherical and keep the old basis stored somewhere so that the user can revert his changes. However, this solution is going to be quite costly in development time.

What I suggest is to have a `get_spherical_components` method that takes a Dataset as the argument but also a dictionary containing all the groups and arrays that the user would like to convert. Something like:

`osyris.spatial.get_spherical_components(data, {"amr":"position", "hydro": ["velocity", "B_field"]})`

It should be easier in this case to make a few checks on the user's request (eg. if they request spherical velocity components, we can access the position array since the whole dataset has been passed as an argument).

Reply:
I forgot to say that this suggestion would return a new Dataset containing all the converted coordinates and `.r`, `.theta` and `.phi` components for each array.

Eg. :
```
spherical_coords = osyris.spatial.get_spherical_components(data, {"amr":"position", "hydro": ["velocity", "B_field"]})

spherical_coords["amr"]["position"].r # radius
spherical_coords["amr"]["position"].theta # colatitude
spherical_coords["amr"]["position"].phi # azimuth

spherical_coords[“hydro"]["velocity"].r # radial velocity
spherical_coords[“hydro"]["velocity"].phi # azimuthal velocity
etc...
```

And `spherical_coords["amr"]["position"].x` returns an attribute error.